lib/maths/CBoostedTreeUtils.cc

```diff
@@ -74,9 +101,109 @@ void writeLossCurvature(const TRowRef& row,
+
+    if (forest.size() < 2) {
+        return {};
+    }
+
+    std::size_t numberLossParameters{loss.numberParameters()};
+    LOG_TRACE(<< "dependent variable " << dependentVariable);
+    LOG_TRACE(<< "number loss parameters = " << numberLossParameters);
+
+    auto makeComputeTotalLossGradient = [&]() {
+        return [&](TDoubleVectorVecVec& leafLossGradients, TRowItr beginRows, TRowItr endRows) {
+            for (auto row = beginRows; row != endRows; ++row) {
+                auto prediction = readPrediction(*row, extraColumns, numberLossParameters);
+                double actual{readActual(*row, dependentVariable)};
+                double weight{readExampleWeight(*row, extraColumns)};
+                for (std::size_t i = 0; i < forest.size(); ++i) {
+                    auto encodedRow = encoder.encode(*row);
+                    std::size_t leaf{root(forest[i]).leafIndex(encodedRow, forest[i])};
+                    auto writer = [&](std::size_t j, double gradient) {
+                        leafLossGradients[i][leaf](j) = gradient;
+                    };
+                    loss.gradient(encodedRow, true, prediction, actual, writer, weight);
+                }
+            }
+        };
+    };
+
+    auto makeZeroLeafLossGradients = [&] {
+        TDoubleVectorVecVec leafLossGradients;
+        leafLossGradients.reserve(forest.size());
+        for (const auto& tree : forest) {
+            leafLossGradients.emplace_back(tree.size(), TDoubleVector::Zero(numberLossParameters));
+        }
+        return leafLossGradients;
+    };
+
+    auto computeLeafLossGradients = [&](const core::CPackedBitVector& rowMask) {
+        auto results = frame.readRows(
+            numberThreads, 0, frame.numberRows(),
+            core::bindRetrievableState(makeComputeTotalLossGradient(),
+                                       makeZeroLeafLossGradients()),
+            &rowMask);
+
+        TDoubleVectorVecVec leafLossGradients{std::move(results.first[0].s_FunctionState)};
+        for (std::size_t i = 1; i < results.first.size(); ++i) {
+            auto& resultsForWorker = results.first[i].s_FunctionState;
+            for (std::size_t j = 0; j < resultsForWorker.size(); ++j) {
+                for (std::size_t k = 0; k < resultsForWorker[j].size(); ++k) {
+                    leafLossGradients[j][k] += resultsForWorker[j][k];
+                }
+            }
+        }
+
+        return leafLossGradients;
+    };
+
+    // Compute the sum of loss gradients for each leaf for the old and
+    // new data.
+
+    auto trainingDataLossGradients = computeLeafLossGradients(trainingDataRowMask);
+
+    for (std::size_t i = 0; i < forest.size(); ++i) {
+        propagateLossGradients(rootIndex(), forest[i], trainingDataLossGradients[i]);
+    }
+    LOG_TRACE(<< "loss gradients = "
+              << core::CContainerPrinter::print(trainingDataLossGradients));
+
+    // We interested in choosing trees for which the total gradient at all
```

Comment:
```suggestion
    // We are interested in choosing trees for which the total gradient at all
```

Reply:
I corrected and also expanded this comment slightly because I think it is a useful reminder. See [ea93475](https://github.com/elastic/ml-cpp/pull/1820/commits/ea934756759014d742ec625bde938a3d8cb24eb0).